java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,296 @@
+  TaintedSpringRequestBody() {
+    exists(SpringServletInputAnnotation a | this.asParameter().getAnAnnotation() = a)
+  }
+}
+
+/**
+ * A sink for Expresssion Language injection vulnerabilities via Jexl,
+ * i.e. method calls that run evaluation of a Jexl expression.
+ *
+ * Creating a `Callable` from a tainted Jexl expression or script is considered as a sink
+ * although the tainted expression is not executed at this point.
+ * Here we assume that it will get executed at some point,
+ * maybe stored in an object field and then reached by a different flow.
+ */
+private class JexlEvaluationSink extends DataFlow::ExprNode {
+  JexlEvaluationSink() {
+    exists(MethodAccess ma, Method m, Expr taintFrom |
+      ma.getMethod() = m and taintFrom = this.asExpr()
+    |
+      m instanceof DirectJexlEvaluationMethod and ma.getQualifier() = taintFrom
+      or
+      m instanceof CreateJexlCallableMethod and ma.getQualifier() = taintFrom
+      or
+      m instanceof JexlEngineGetSetPropertyMethod and
+      taintFrom.getType() instanceof TypeString and
+      ma.getAnArgument() = taintFrom
+    )
+  }
+}
+
+/**
+ * Defines method calls that propagate tainted data via one of the methods
+ * from Jexl library.
+ */
+private class TaintPropagatingJexlMethodCall extends MethodAccess {
+  Expr taintFromExpr;
+
+  TaintPropagatingJexlMethodCall() {
+    exists(Method m, RefType taintType |
+      this.getMethod() = m and
+      taintType = taintFromExpr.getType()
+    |
+      isUnsafeEngine(this.getQualifier()) and
+      (
+        m instanceof CreateJexlScriptMethod and
+        taintFromExpr = this.getArgument(0) and
+        taintType instanceof TypeString
+        or
+        m instanceof CreateJexlExpressionMethod and
+        taintFromExpr = this.getAnArgument() and
+        taintType instanceof TypeString
+        or
+        m instanceof CreateJexlTemplateMethod and
+        (taintType instanceof TypeString or taintType instanceof Reader) and
+        taintFromExpr = this.getArgument([0, 1])
+      )
+    )
+  }
+
+  /**
+   * Holds if `fromNode` to `toNode` is a dataflow step that propagates
+   * tainted data.
+   */
+  predicate taintFlow(DataFlow::Node fromNode, DataFlow::Node toNode) {
+    fromNode.asExpr() = taintFromExpr and toNode.asExpr() = this
+  }
+}
+
+/**
+ * Holds if `expr` is a Jexl engine that is not configured with a sandbox.
+ */
+private predicate isUnsafeEngine(Expr expr) {
+  not exists(SandboxedJexlFlowConfig config | config.hasFlowTo(DataFlow::exprNode(expr)))
+}
+
+/**
+ * A configuration for a tracking sandboxed Jexl engines.
+ */
+private class SandboxedJexlFlowConfig extends DataFlow2::Configuration {
+  SandboxedJexlFlowConfig() { this = "JexlInjection::SandboxedJexlFlowConfig" }
+
+  override predicate isSource(DataFlow::Node node) { node instanceof SandboxedJexlSource }
+
+  override predicate isSink(DataFlow::Node node) {
+    node.asExpr().getType() instanceof JexlEngine or
```

Comment:
This could produce a pretty large set of sinks. Looks like we're only interested in qualifiers to CreateJexlScriptMethod calls and similar -- suggest restricting this to only consider the arguments we will consider in TaintPropagatingJexlMethodCall's characteristic predicate.

Reply:
I felt that something is wrong with this sink but was not sure. I thought if I added `CreateJexlScriptMethod` and similar calls here, it would be a bit redundant because they are already covered in `TaintPropagatingJexlMethodCall`. I've updated the predicate to consider the methods covered in `TaintPropagatingJexlMethodCall`. Thanks for the suggestion!